View Skipped Days In The Bar Chart

Mahmoud-Ibrahim-750 opened the pull request:
I noticed that the skipped days appear in the history chart but not in the bar chart so I modified the code adding a piece of code to pass the skipped days data from the database to the view and drawing a bar with a slightly transparent colour -to let it be different from the main column- accordingly.

Mahmoud-Ibrahim-750 committed:
View Skipped Days In The Bar Chart

I noticed that the skipped days appear in the history chart but not in the bar chart so I modified the code adding a piece of code to pass the skipped days data from database to the view and drawing a bar with a slightly transparent color -to let it be different from the main column- accordingly.

Mahmoud-Ibrahim-750 committed:
Merge branch 'iSoron:dev' into dev

Mahmoud-Ibrahim-750 commented on uhabits-android/src/main/java/org/isoron/uhabits/activities/habits/show/views/BarCardView.kt:
```diff
@@ -40,6 +40,7 @@ class BarCardView(context: Context, attrs: AttributeSet) : LinearLayout(context,
         val androidColor = state.theme.color(state.color).toInt()
         binding.chart.view = BarChart(state.theme, JavaLocalDateFormatter(Locale.US)).apply {
             series = mutableListOf(state.entries.map { it.value / 1000.0 })
+            skippedSeries = mutableListOf(state.skippedEntries.map { it.value / 1.0 })
```
pass the skipped days to the view

Mahmoud-Ibrahim-750 committed:
Update BarChartTest.kt

Updated the tests to provide the data for skipped entries too.

hiqua commented on uhabits-core/src/jvmTest/java/org/isoron/uhabits/core/ui/views/BarChartTest.kt:
```diff
@@ -34,10 +34,12 @@ class BarChartTest {
     init {
         component.axis = axis
         component.series.add(series1)
+        component.skippedSeries.add(series2)
```
How come the test setup seems to change, but the assertions (the .png) do not? Are they just close enough such that the tests pass anyway, and should they be [updated](https://github.com/iSoron/uhabits/blob/dev/docs/TEST.md) instead? What am I missing?

Mahmoud-Ibrahim-750 replied:
After running the test multiple times with different cases, here is a final conclusion:
First, the test mechanism seems to ignore little differences between the expected image and the actual one, so yes, they are close enough such that the test pass as long as the difference value is less than 1. 
Second, I think the test mechanism should be updated to take care of little difference, then according to that update, the expected image should be updated too. We may update it for now until we take care of the test in a different PR.

Here are some of the results I got from the test, feel free to try them on your own (each test case is provided with the difference value and the output image too):
1. the current expected image (the one we may update later):
[[Imgur](https://i.imgur.com/59dDEHg.png)]

2. the current test data (with skipped days):
[[Imgur](https://i.imgur.com/1oPJApQ.png)](current test data with skipped days)
[[Imgur](https://i.imgur.com/8dAKgMb.png)](output)
[[Imgur](https://i.imgur.com/dPUrIMS.png)](difference)

3. the current test data (without skipped days):
[[Imgur](https://i.imgur.com/gw1xmMD.png)](current test data without skipped days)
[[Imgur](https://i.imgur.com/NyBa4Aw.png)](output)
[[Imgur](https://i.imgur.com/u3xCRoI.png)](difference)

4. a modified test data (accepted):
[[Imgur](https://i.imgur.com/2dxzoZm.png)](accepted modified test data)
[[Imgur](https://i.imgur.com/DbUWvT4.png)](output)
[[Imgur](https://i.imgur.com/zt9LtHH.png)](difference)

5. a modified test data (rejected):
[[Imgur](https://i.imgur.com/Oq74hpB.png)](rejected modified test data)
[[Imgur](https://i.imgur.com/rvRDqPu.png)](output)
[[Imgur](https://i.imgur.com/thrNato.png)](difference)

hiqua replied:
I think it's better to update the expected images in the same PR, following https://github.com/iSoron/uhabits/blob/dev/docs/TEST.md#running-instrumented-tests, but you can also wait for @iSoron to chime in if you prefer.

Mahmoud-Ibrahim-750 replied:
It's ok to wait. Updating the image takes no time anyway, So there's no need to rush.

Mahmoud-Ibrahim-750 replied:
@iSoron, Would you check this PR so we can proceed and finally merge it?

Mahmoud-Ibrahim-750 replied:
@hiqua Test image is up-to-date and this PR is ready for a final review.

Mahmoud-Ibrahim-750 committed:
Update the expected image

I updated the expected image in the test to take into account the skipped days as well as the completed days. I also changed the test input in BarChartTest.kt to let it better represent possible cases.
